mysql/datadog_checks/mysql/execution_plans.py

```diff
@@ -307,32 +320,129 @@ def _collect_execution_plans(self, db, tags, options, min_collection_interval):
         statsd.gauge("dd.mysql.collect_execution_plans.seen_statement_plan_sigs", len(seen_statement_plan_sigs),
                      tags=instance_tags)
 
-    def _run_explain(self, cursor, statement, schema, instance_tags):
-        # TODO: cleaner query cleaning to strip comments, etc.
-        if statement.strip().split(' ', 1)[0].lower() not in VALID_EXPLAIN_STATEMENTS:
-            return
+    def _attempt_explain(self, cursor, statement, schema):
+        """
+        Tries the available methods used to explain a statement for the given schema. If a non-retryable
+        error occurs (such as a permissions error), then statements executed under the schema will be
+        disallowed in future attempts.
+        """
+        explain_strategy_none = 'none'
+        explain_strategy_procedure = 'procedure'
+        explain_strategy_statement = 'statement'
+
+        fns_by_strategy = {
+            explain_strategy_procedure: self._run_explain_procedure,
+            explain_strategy_statement: self._run_explain,
+        }
+
+        plan = None
+        strategy_cache_key = 'explain_strategy:%s' % schema
+        strategy_cache_ttl = 60 * 10
 
+        if not self._can_explain(statement):
+            return None
+
+        if self._expiring_cache.get(strategy_cache_key) == explain_strategy_none:
+            return None
+
+        # Switch to the right schema
+        try:
+            self._use_schema(cursor, schema)
+        except NonRetryableError:
+            self._expiring_cache.set(strategy_cache_key, explain_strategy_none, strategy_cache_ttl)
+            return None
+        except RetryableError:
+            return None
+
+        # Use a cached strategy for the schema, if any, or try each strategy to collect
+        strategy = self._expiring_cache.get(strategy_cache_key)
+        fn = fns_by_strategy.get(strategy)
+        if fn:
+            plan = fn(cursor, statement)
```

Comment:
`fn` could throw an exception here. We're catching them in the other branch of this flow though, so not sure if the intention is to let it bubble up or catch it here and return none?

Reply:
We should also find some way to consistently log a single warning for each failed plan. We shouldn't require users to enable debug logging in order to see the warnings, if the warnings would help them figure out what's causing the problem. 

Especially all schema, permission, stored procedure, related errors.

Reply:
Yeah the idea is that if we get here, it has already been successful once so any error is actually unexpected. Let me see if the flow can be improved so that there's a single point of logging if the collection fails.